File: android/app/src/main/java/dev/bridgey/android/VideoTransport.kt
package dev.bridgey.android

// Minimal transport abstraction (frozen spec, section 3): expresses transport semantics, not TCP
// specifics. TcpVideoTransport is implementation #1; a future UDP/QUIC implementation satisfies the
// same interface without VideoChannelManager, encoder, or decoder ever changing.

internal data class TransportCapabilities(
    val reliable: Boolean,
    val ordered: Boolean,
    val maxFrameBytes: Int,
)

internal data class TransportMetrics(
    val framesSent: Long = 0,
    val framesReceived: Long = 0,
    val framesDropped: Long = 0,
    val bytesSent: Long = 0,
)

internal sealed interface SendOutcome {
    data object Sent : SendOutcome
    data object QueuedBounded : SendOutcome
    data class Failed(val reason: String) : SendOutcome
}

internal data class EncodedVideoFrame(
    val type: Int, // VideoFrameType.*
    val streamId: Long,
    val captureTimestampMs: Long,
    val payload: ByteArray,
)

internal interface VideoTransport {
    val capabilities: TransportCapabilities
    val metrics: TransportMetrics
    var onFrameReceived: (EncodedVideoFrame) -> Unit
    var onDisconnected: (Throwable?) -> Unit

    /** Enqueues a frame for sending. `droppable=false` (KEYFRAME, CONFIG) must never be silently
     * dropped by backpressure; `droppable=true` (DELTA) may be, per the frozen backpressure policy. */
    fun send(frame: EncodedVideoFrame, droppable: Boolean): SendOutcome
    fun close()
}

File: macos/Sources/BridgeyMac/VideoTransport.swift
import Foundation

// Minimal transport abstraction (frozen spec, section 3): expresses transport semantics, not TCP
// specifics. TCPVideoTransport is implementation #1; a future UDP/QUIC implementation satisfies the
// same protocol without VideoChannelController, encoder, or decoder ever changing.
// Direct port of VideoTransport.kt.

struct TransportCapabilities {
    let reliable: Bool
    let ordered: Bool
    let maxFrameBytes: Int
}

struct TransportMetrics {
    var framesSent: Int64 = 0
    var framesReceived: Int64 = 0
    var framesDropped: Int64 = 0
    var bytesSent: Int64 = 0
}

enum SendOutcome: Equatable {
    case sent
    case queuedBounded
    case failed(String)
}

struct EncodedVideoFrame {
    let type: Int // VideoFrameType.*
    let streamId: Int64
    let captureTimestampMs: Int64
    let payload: Data
}

protocol VideoTransport: AnyObject {
    var capabilities: TransportCapabilities { get }
    var metrics: TransportMetrics { get }
    var onFrameReceived: (EncodedVideoFrame) -> Void { get set }
    var onDisconnected: (Error?) -> Void { get set }

    /// Enqueues a frame for sending. `droppable=false` (KEYFRAME, CONFIG) must never be silently
    /// dropped by backpressure; `droppable=true` (DELTA) may be, per the frozen backpressure policy.
    func send(_ frame: EncodedVideoFrame, droppable: Bool) -> SendOutcome
    func close()
}
